File: exercises/valid-palindrome-125.go
package main

import (
    "unicode"
)
func validPalindrome(s string) bool {
    str := filterString([]byte(s))
    head := 0
    tail := len(str) - 1
    isPalindrome := true
    for head < tail {
        if unicode.ToUpper(rune(str[head])) != unicode.ToUpper(rune(str[tail])) {
            isPalindrome = false
            break
        }
        head += 1
        tail -= 1
    }
    return isPalindrome
}

func isAlphanumeric (c byte) bool {
    // return unicode.IsDigit(c) || unicode.IsLetter(c)
    return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' ||
             c >= '0' && c <= '9'
}

func filterString(str []byte) (ret []byte) {
    for _, s := range str {
        if isAlphanumeric(s) {
            ret = append(ret, s)
        }
    }
    return
}

File: exercises/validPalindrome125.java

public class validPalindrome125 {
    public static void main(String[] args) {
        // var sol = new Solution();
        // var res = sol.isPalindrome("racecar");
        // out.printf("The result is %s%n", res);
    }
}

class Solution_125 {
    public boolean isPalindrome(String s) {
        var str = filterString(s);
        var head = 0;
        var tail = str.length() - 1;
        var isPalindrome = true;

        while (head < tail) {
            if ( Character.toUpperCase(str.charAt(head)) != 
            Character.toUpperCase(str.charAt(tail)) ) {
                isPalindrome = false;
                break;
            }
            head++;
            tail--;
        }
        return isPalindrome;
    }

    static String filterString (String str) {
        var result = new StringBuilder();
        for (char ch : str.toCharArray()) {
            if (Character.isLetterOrDigit(ch)) {
                result.append(ch);
            }
        }
        return result.toString();
    }
}
